SQL changes workover

a-tze opened the pull request:
See issue #166 .

Functionality testing still outstanding. Some minor, mostly cosmetic differences to current maser exist.

Feel free to add commits to this branch, I'll squash them in before removing WIP status. Accordingly, expect history rewrites in this branch as long as it is in WIP status.

The current last migration file will be replaced in a final commit when removing WIP state. If there are other migrations in the meantime, then a suitable change will be added.

pegro commented:
First of all: thanks for (re-)working (on) this.
Second: I would prefer keeping all these commits separate in the history and don't squash them. I think we aren't really bisectable anyway, so that shouldn't be a concern.

I'll try to comment on the other things inline.

pegro commented on src/Application/Migrations/06_tickets.sql:
```diff
@@ -136,6 +136,9 @@ WITHOUT OIDS;
 CREATE INDEX tbl_ticket_handle_id_idx ON tbl_ticket USING btree(handle_id);
+CREATE INDEX tbl_ticket_parent_id_idx ON tbl_ticket USING hash(parent_id);
+CREATE INDEX tbl_ticket_project_id_idx ON tbl_ticket USING hash(project_id);
+CREATE INDEX tbl_ticket_view_servicable_idx ON tbl_ticket USING btree(failed, service_executable, ticket_type);
```
Mh I guess the whole migration patch file is not updated.

jjeising commented on src/Application/Migrations/17_function_ticket_dependency_satisfied.sql:
```diff
@@ -27,11 +27,20 @@ BEGIN
+	JOIN
+		tbl_ticket_state configured_trigger_state ON
+			t.ticket_type = 'encoding' AND
+			configured_trigger_state.ticket_state = p.dependent_ticket_trigger_state
```
> or just use "master ticket", whenever you mean the master ticket.

I'm probably the reason we're not currently doing this. As we set `depends_on` for encoding profiles, we have no notion of a _master_ or an informal _subticket_ anywhere else in the UI or code. Therefor i suggested something like _dependent_. I like _dependee_ and maybe _dependent_? The last is a little more common. This would also free us from the language issues around _master/slave_.

jjeising replied:
>  in #171 it's about to be called that, if I saw that correctly.

I think this is only a comment and a UI label – I already requested changing this some time ago.

a-tze replied:
The semantical change of the parameter is likely an error, I will investigate that.

a-tze commented on src/Application/Migrations/17_function_ticket_dependency_satisfied.sql:
```diff
@@ -27,11 +27,20 @@ BEGIN
 
 	SELECT
-		state >= p.dependent_ticket_trigger_state INTO satisfaction
+		dependent_ticket_state.sort >= configured_trigger_state.sort INTO satisfaction
```
Not sure if this breaks the possibility of later changes without deleting all tickets, e.g. not being able to REPLACE the enum when there are references to it. 
It is quite some code to use the sort property, but may be better in the long run.

pegro replied:
Inplace replacing is difficult, that's true. But one could introduce the new enum with a temporary name, add temporary columns to all related tables, copy the value from the old column, remove the old one and rename the enum column back to the old name. Should™ work ;)

pegro replied:
Okay, see #173

a-tze commented:
@pegro The "squashing in" did not mean I want to squash the commits/ the PR once it is approved. It's the solely purpose of this PR to introduce multiple commits with more explanation.
What I would like to squash into other commits are the corrections that are done now, e.g. renaming the duplicate index or stuff like that, just to have a super clean history.
Does that make sense?

a-tze commented on src/Application/Controller/XMLRPC/Handler.php:
```diff
@@ -616,7 +616,7 @@ public function assignNextUnassignedForState($ticketType = '', $ticketState = ''
 						'virtual_property_filter' => [$propertyFilters]
 					])
-					->orderBy('ticket_priority(id) DESC');
+					->orderBy('calculated_priority DESC');
```
This is probably just legacy and/or a case of "better safe than sorry".

a-tze commented on src/Application/Migrations/22_view_serviceable_tickets.sql:
```diff
@@ -36,7 +36,7 @@ CREATE OR REPLACE VIEW view_serviceable_tickets AS
 	WHERE
 		pj.read_only = false AND
-		t.ticket_type != 'meta' AND
+		t.ticket_type IN ('recording','encoding','ingest') AND
```
Because for some reasons Postgres doesnt get it that this is actually the same. The newer line will result in using the ANY operation which will in turn make use of a hash index, while using != will result in a "simple filter" on a full index scan or even table scan.

pegro commented:
Yes, makes sense. ;)

a-tze commented:
Ok.. rebased to current master, added migration and did some testing. The branch now includes a somewhat consistent renaming to dependee and depender.

@pegro @jjeising please do a final review. Note: Github seems to get the order of commits wrong.

a-tze commented on src/Application/Migrations/06_tickets.sql:
```diff
@@ -137,7 +137,6 @@ CREATE INDEX tbl_ticket_project_id_idx ON tbl_ticket USING btree(project_id);
 CREATE INDEX tbl_ticket_fahrplan_id_idx ON tbl_ticket USING btree(fahrplan_id);
 CREATE INDEX tbl_ticket_handle_id_idx ON tbl_ticket USING btree(handle_id);
 CREATE INDEX tbl_ticket_parent_id_idx ON tbl_ticket USING hash(parent_id);
```
This is intentional. You usually do not query "less than" or "greater than" for things like IDs. Unfortunately in my tests those indexes were not used according to the execution plan when they were of type btree. Makes some sense in my head for JOINS. I can test that again if you think btree should be used by the query optimizer in every case a hash index is used.

a-tze commented on src/Application/Migrations/17_function_ticket_dependency_satisfied.sql:
```diff
@@ -2,39 +2,33 @@
-			wantedstate.ticket_type = t2.ticket_type AND
-			wantedstate.ticket_state = p.dependent_ticket_trigger_state
+		tbl_ticket_state configured_trigger_state ON
+			depender_ticket.ticket_type = 'encoding' AND
```
But ticket_state is an enum and therefore unique, isnt it? The JOIN is meant to deliver the state attributes for encoding tickets and NULLs for other ticket types. So is your concern that these JOINS produce more tupels? I cannot think of a case where this might happen.

If your concern is that dependee_ticket_state and configured_trigger_state might contain the same tupel of the state table, then yes this is possible of course.

pegro replied:
ticket_state is an enum, but tbl_ticket_state is a table (which you are joining here) and it contains multiple rows with e.g. ```ticket_state = 'removing'```. I know, it's not really likely that the trigger state would be one of those, but technically it could happen.

> The JOIN is meant to deliver the state attributes for **encoding** tickets
But the ticket_type of **tbl_ticket_state** is not part of any condition... That's what I'm referring to.

pegro replied:
Looks good, although I'd have used ```dependee_ticket_state.ticket_type = depender_ticket.ticket_type``` just to have the hard-coded value only once.

pegro replied:
And since you don't do any left joins and such, I'd suggest to move the ```depender_ticket.ticket_type = 'encoding'``` condition to the where clause, to only have it once. The function should still return NULL, if the ticket is not a encoding ticket.

pegro commented on src/Application/Migrations/__2018-04-15_add_variable_dependent_ticket_state.sql:
```diff
@@ -3,17 +3,17 @@ BEGIN;
 SET ROLE TO postgres;
```
Yes, I'd be in favor of shrinking.

a-tze commented:
@pegro please have a look at the last 3 commits from the latest batch. Another round of testing is still outstanding.

pegro commented on src/Application/Migrations/22_view_serviceable_tickets.sql:
```diff
@@ -34,40 +30,46 @@ CREATE OR REPLACE VIEW view_serviceable_tickets AS
-	LEFT JOIN
-		tbl_ticket_state wantedstate ON wantedstate.ticket_type = 'encoding' AND wantedstate.ticket_state = pj.dependent_ticket_trigger_state
+		tbl_ticket_state configured_trigger_state ON
+			configured_trigger_state.ticket_type = 'encoding' AND
```
I guess ```configured_trigger_state.ticket_type = t.ticket_type``` would do the same, but it always be _encoding_ as long we don't allow other ticket types as ```encoding``` to have a _encoding_profile_version_id_ set.
I'm ok with this. It's just I like to avoid as much hard-coding as possible without performance impact.

a-tze commented:
@pegro good idea, I made this change for the function. Ommitting the change for the view right now.

Also @jjeising: If no objections exist, I will merge this in the next 3 days so that the other PRs can move forward.

pegro commented:
Fine with me
